x-pack/plugins/ml/server/lib/route_guard.ts

```diff
@@ -21,14 +21,14 @@ import { MlLicense } from '../../common/license';
 
 import { MlClient, getMlClient } from '../lib/ml_client';
 
-type Handler = (handlerParams: {
+type Handler<P = unknown, Q = unknown, B = unknown> = (handlerParams: {
```

Comment:
Core provides a type-safe definition for request handler wrapper https://github.com/elastic/kibana/blob/5342877a32b213ef18cef2fcf23ea4f549f30e65/src/core/server/http/router/router.ts#L376

Reply:
Yea, but our wrapper makes the assumption that the lower-order 'handler' has the exact same signature as the vanilla route handler, which is not the case here, so I don't think we can directly use it?

Reply:
Can we provide another one with a variadic number of arguments then?

Reply:
With a variable number of arguments, probably yea. With a single argument containing the handler's arguments, as in this specific example, I don't think so:

```ts
type Handler = (handlerParams: {
  client: IScopedClusterClient;
  request: KibanaRequest<any, any, any, any>;
  response: KibanaResponseFactory;
  context: RequestHandlerContext;
  jobSavedObjectService: JobSavedObjectService;
  mlClient: MlClient;
}) => ReturnType<RequestHandler>;
```

if `Handler` was to be changed to something like

```ts
type Handler = (
  context: RequestHandlerContext,
  request: KibanaRequest<any, any, any, any>,
  response: KibanaResponseFactory,
  { client, jobSavedObjectService, mlClient }: {
     client: IScopedClusterClient;
     jobSavedObjectService: JobSavedObjectService;
     mlClient: MlClient;
  } 
}) => ReturnType<RequestHandler>;
```

it would probably be more doable. Now, that would requires changes in current consumers code